### PR TITLE
Adds new join capabilities - WIP

### PR DESCRIPTION
**Pull request by systay**
I've tried keeping the commits as small as possible. We can always squash before merging.

### CODE REVIEW

File: go/vt/vtgate/engine/hash_join.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+		p.table[hash] = [][]sqltypes.Value{value}
+	} else {
+		p.table[hash] = append(chunk, value)
+	}
+}
+
+func (p *probeTable) calculateHashFor(key []sqltypes.Value) int {
+	hash := 0
+	for _, k := range key {
+		hash = 31*hash + int(k.Type())
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/engine/hash_join.go]: Don't know what it's called. It's the standard algorithm used by Java: See http://developer.classpath.org/doc/java/util/Arrays-source.html, line 2669

We are relying on the go hash-map - look at line 186 in this file. The question is if we should turn the slice into an int or a string to use as a key, right?

File: go/vt/vtgate/engine/hash_join.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	// LeftJoinCols defines which columns from the lhs are part of the ON comparison
+	LeftJoinCols []int
+
+	// RightJoinCols defines which columns from the rhs are part of the ON comparison
+	RightJoinCols []int
+}
+
+// MarshalJSON allows us to add the opcode in, so we can see the join type used
+func (jn *HashJoin) MarshalJSON() ([]byte, error) {
+	type Alias HashJoin
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/engine/hash_join.go]: I added this to see in the query plan output what the type of join it is. Not sure how to do that without adding a fake field to the struct, which seemed worse to me.

File: go/vt/vtgate/engine/hash_join.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+type probeTable struct {
+	table map[int][][]sqltypes.Value
+}
+
+func newProbeTable() probeTable {
+	return probeTable{table: make(map[int][][]sqltypes.Value)}
+}
+
+func (p *probeTable) Add(key []sqltypes.Value, value []sqltypes.Value) {
+	hash := p.calculateHashFor(key)
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/engine/hash_join.go]: ah, so if the user asks for a join on a text column, like `ON t1.varchar = t2.varchar1`, we can turn it into a `ON WEIGHT_STRING(t1.varchar) = WEIGHT_STRING (t2.varchar1)` and perform the hashing on that comparison instead?

File: go/vt/vtgate/engine/hash_join.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+			if resultFields == nil {
+				resultFields = joinFields(leftFields, rresult.Fields, jn.Cols)
+			}
+			result.Fields = resultFields
+		}
+
+		for _, rrow := range rresult.Rows {
+			joinVals := extractJoinValues(rrow, jn.RightJoinCols)
+			matches := table.Get(joinVals)
+			for _, lrow := range matches {
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/engine/hash_join.go]: Good catch. I'll put that logic into the probe table.

File: go/vt/vtgate/engine/hash_join.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/engine/hash_join.go]: Not sure I understand the comment about the type. Do you mean to say that "1" should be joined with int32(1)?

File: go/vt/vtgate/engine/hash_join.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vtgate/engine/hash_join.go]: Something like this?

```
	return json.Marshal(&struct {
		OpCode string `json:"Opcode"`
		*HashJoin
	}{
		OpCode: "HashJoin",
		HashJoin:  jn,
	})
```

File: go/vt/vtgate/engine/hash_join.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vtgate/engine/hash_join.go]: I was thinking we should construct a combined key as concatenation of the string representation of the keys, and just give it to go's map, which is already heavily optimized. And if there's a match, we do the individual compare to confirm.

File: go/vt/vtgate/engine/hash_join.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/vtgate/engine/hash_join.go]: nvm. I see the problem. It becomes recursive this way.

File: go/vt/vtgate/planbuilder/join.go [hunk 5]
```diff
@@ -150,22 +152,53 @@ func createJoinStruct(lpb, rpb *primitiveBuilder, opcode engine.JoinOpcode, useH
...
+		if err != nil {
+			return nil, err
+		}
+		rgt, err := extractJoinComparisons(and.Right)
+		if err != nil {
+			return nil, err
+		}
+		return append(lft, rgt...), nil
+	}
+	return nil, vterrors.Errorf(vtrpc.Code_INTERNAL, "can only handle simple equality between columns using hash joins - got %s", expr)
```
Comment (mpawliszyn) [inline go/vt/vtgate/planbuilder/join.go]: I think you might want to express that it must be a list of ands in this case?

File: go/vt/vtgate/planbuilder/join.go [hunk 6]
```diff
@@ -98,22 +111,126 @@ func newJoin(lpb, rpb *primitiveBuilder, ajoin *sqlparser.JoinTableExpr) error {
...
+		if err != nil {
+			return nil, err
+		}
+		rgt, err := extractJoinComparisons(and.Right)
+		if err != nil {
+			return nil, err
+		}
+		return append(lft, rgt...), nil
+	}
+	return nil, vterrors.Errorf(vtrpc.Code_INTERNAL, "can only handle one or more equality comparisons between columns using hash joins - got %s", expr)
```
Comment (mpawliszyn) [inline go/vt/vtgate/planbuilder/join.go]: I think you should add the fact that all the comparisons must be conjunctive(a bunch of and experessions).

**Comment by systay** [comment]
Fun experiment. Not ready for prime time

**Pull request closed without merging**